src/ngraph/op/fused/depth_to_space.cpp

```diff
@@ -46,59 +47,100 @@ op::DepthToSpace::DepthToSpace(const Output<Node>& data,
 NodeVector op::DepthToSpace::decompose_op() const
 {
     auto data = input_value(0);
-    const Shape& data_shape = data.get_shape();
+    auto data_shape = data.get_shape();
 
-    // Set default values to each dimension to be able to work with both 3D or 4D data.
-    size_t n{1}, c{1}, h{1}, w{1};
+    NGRAPH_CHECK((data_shape.size() >= 3),
+                 "The input tensor with rank lower than 3 is not supported (input rank: ",
+                 data_shape.size(),
+                 ")");
 
-    NGRAPH_CHECK((data_shape.size() == 3 || data_shape.size() == 4),
-                 "The provided tensor shape: ",
-                 data_shape,
-                 " is not supported.");
-
-    // Assume NCHW data layout
-    if (data_shape.size() == 4)
-    {
-        n = data_shape.at(0);
-        c = data_shape.at(1);
-        h = data_shape.at(2);
-        w = data_shape.at(3);
-    }
-    // Without batch.
-    else if (data_shape.size() == 3)
+    if (data_shape.size() == 3)
     {
-        c = data_shape.at(0);
-        h = data_shape.at(1);
-        w = data_shape.at(2);
+        // Insert batch axis
+        data_shape.insert(data_shape.begin(), 1);
+        data = builder::reshape(data, data_shape);
     }
-
-    NGRAPH_CHECK((c % (m_blocksize * m_blocksize) == 0 && m_blocksize > 0),
-                 "SpaceToDepth: The depth axis size must be a multiple of ",
-                 "squared block_size attribute value.");
+    const size_t n_dim = data_shape.at(0);
+    const size_t c_dim = data_shape.at(1);
+    const size_t spatial_dims = data_shape.size() - 2;
+    const auto c_dim_divider = static_cast<int>(std::pow(m_blocksize, spatial_dims));
+
+    NGRAPH_CHECK(m_blocksize > 0 && c_dim % c_dim_divider == 0,
+                 "The color axis size: ",
+                 c_dim,
+                 " must be a multiple of ",
+                 "block_size^spatial_dims: ",
+                 c_dim_divider,
+                 " attribute value");
 
     auto bs = static_cast<size_t>(m_blocksize);
-    size_t c_flat = c / (bs * bs);
+    size_t c_flat = c_dim / c_dim_divider;
 
     // First we have to disperse the data from depth channel, then rearrange them
     // so as appropriate chunks of data where close to their destination place.
     // Finally squeeze data from respective dimensions.
     shared_ptr<Node> flat_node;
+    Shape dispersed_shape{n_dim};
+    for (int i = 0; i < spatial_dims; ++i)
+    {
+        dispersed_shape.push_back(bs);
+    }
+    const auto no_spatial_dims = data_shape.size() - spatial_dims;
```

Comment:
I'm not a fan of this variable please remove/rename it.
This variable will be always equal to `2`.

Reply:
I used `spatial_dim_index`